modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Enrich.scala

```diff
@@ -0,0 +1,132 @@
+ */
+package com.snowplowanalytics.snowplow.enrich.fs2
+
+import java.time.Instant
+import java.util.Base64
+import java.util.concurrent.TimeUnit
+
+import org.joda.time.DateTime
+
+import cats.data.{NonEmptyList, ValidatedNel}
+import cats.implicits._
+
+import cats.effect.{Clock, Concurrent, ContextShift, Sync}
+
+import fs2.Stream
+
+import _root_.io.sentry.SentryClient
+
+import _root_.io.circe.Json
+import _root_.io.circe.syntax._
+
+import _root_.io.chrisdavenport.log4cats.Logger
+import _root_.io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+
+import com.snowplowanalytics.iglu.client.Client
+
+import com.snowplowanalytics.snowplow.badrows.{Processor, BadRow, Failure, Payload => BadRowPayload}
+
+import com.snowplowanalytics.snowplow.enrich.common.EtlPipeline
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+import com.snowplowanalytics.snowplow.enrich.common.adapters.AdapterRegistry
+import com.snowplowanalytics.snowplow.enrich.common.loaders.{CollectorPayload, ThriftLoader}
+import com.snowplowanalytics.snowplow.enrich.common.enrichments.EnrichmentRegistry
+
+object Enrich {
+
+  /** Default adapter registry, can be constructed dynamically in future */
+  val adapterRegistry = new AdapterRegistry()
+
+  val processor: Processor = Processor(generated.BuildInfo.name, generated.BuildInfo.version)
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  /**
+   * Run a primary enrichment stream, reading from [[Environment]] source, enriching
+   * via [[enrichWith]] and sinking into [[GoodSink]] and [[BadSink]] respectively.
+   * Can be stopped via _stop signal_ from [[Environment]]
+   */
+  def run[F[_]: Concurrent: ContextShift: Clock](env: Environment[F]): Stream[F, Unit] = {
+    val enrich: Enrich[F] = enrichWith[F](env.enrichments.registry, env.resolver, env.sentry)
+    env.source
+      .pauseWhen(env.stop)
+      .parEvalMapUnordered(16)(payload => env.blocker.blockOn(enrich(payload)))
+      .flatMap(_.decompose[BadRow, EnrichedEvent])
+      .observeEither(env.bad, env.good)
+      .void
+  }
+
+  /** Enrich a single [[CollectorPayload]] to get list of bad rows and/or enriched events */
+  def enrichWith[F[_]: Clock: Sync](
+    enrichRegistry: EnrichmentRegistry[F],
+    igluClient: Client[F, Json],
+    sentry: Option[SentryClient]
+  )(
+    row: Payload[F, Array[Byte]]
+  ): F[Result[F]] = {
+    val payload = ThriftLoader.toCollectorPayload(row.data, processor)
+    val result = Logger[F].debug(payloadToString(payload)) *>
+      Clock[F]
+        .realTime(TimeUnit.MILLISECONDS)
+        .map(millis => new DateTime(millis))
+        .flatMap { now =>
+          EtlPipeline.processEvents[F](adapterRegistry, enrichRegistry, igluClient, processor, now, payload)
+        }
+        .map(enriched => Payload(enriched, row.ack))
+
+    result.handleErrorWith(sendToSentry[F](row, sentry))
```

Comment:
Shouldn't we send the `payload` rather than the `row` ? So that to troubleshoot we don't need to Thrift deserialize it

Reply:
I think `sendToSentry` is probably slightly misleading. It does three things:

1. Sends an exception to Sentry (we cannot send anything from an event because it can contain PII data
2. Creates a `generic_error` bad row - that's why we need a `row`
3. Logs an error

Reply:
My point is, why not use `payload` instead of `row` in the generic bad row created? An array of Thrift bytes is not very useful to troubleshoot, compared with a `BadRow` (CPFormatViolation) or a `CollectorPayload`

Reply:
Ah, ok makes sense now!

The reason was that our payload is something like `ValidatedNel[CPFormatViolation, Option[CollectorPayload]]`, so we don't really have a parsed payload yet. We technicall can pattern-match on it, like:

```scala
payload match {
  case Validated.Invalid(errors) => // what to do here? why we were trying to process error in a first place
    errors // thrift bytes anyway
  case Validated.Valid(payload) =>
    turnIntoAdapterFailure(payload) // but what if it's not an adapter failure? enrichment failure? then we need a raw event to construct bad row
```

And also feels weird to produce different kinds of bad row from the same place, so I decided to stick with the most generic one.

Reply:
I was thinking of putting something like `show"$payload"` in the `generic_error`, whatever it contains.

Reply:
It's actually a [bug](https://github.com/snowplow/enrich/issues/56) that `ThriftLoader.toCollectorPayload`'s signature says that it can return multiple bad rows, it can return only one.

So in the `generic_error` bad row we could put either the single `CPFormatViolation` (recoverable) or the raw extracted `CollectorPayload` (recoverable).